File: apps/mobile/lib/services/auth_service.dart
import '../core/device/device_id.dart';
import '../core/network/api_client.dart';
import '../core/storage/secure_store.dart';

class AuthSession {
  AuthSession({
    required this.userId,
    required this.email,
    required this.accessToken,
    required this.refreshToken,
    required this.isVerified,
  });

  final String userId;
  final String email;
  final String accessToken;
  final String refreshToken;
  final bool isVerified;

  factory AuthSession.fromJson(Map<String, dynamic> json) {
    final user = json['user'] as Map<String, dynamic>? ?? {};
    return AuthSession(
      userId: user['id']?.toString() ?? '',
      email: user['email']?.toString() ?? '',
      accessToken: json['accessToken']?.toString() ?? '',
      refreshToken: json['refreshToken']?.toString() ?? '',
      isVerified: user['isVerified'] == true,
    );
  }
}

class AuthService {
  factory AuthService({SecureStore? store}) {
    final resolved = store ?? SecureStore();
    return AuthService._(
      resolved,
      DeviceIdStore(resolved),
      ApiClient(resolved),
    );
  }

  AuthService._(this._store, this._deviceIdStore, this._client);

  final SecureStore _store;
  final DeviceIdStore _deviceIdStore;
  final ApiClient _client;

  Future<AuthSession> login({
    required String email,
    required String password,
  }) async {
    final deviceId = await _deviceIdStore.getOrCreate();
    final data = await _client.post('/auth/login', body: {
      'email': email,
      'password': password,
      'deviceId': deviceId,
    });

    final session = AuthSession.fromJson(data);
    await _saveSession(session);
    return session;
  }

  Future<AuthSession> register({
    required String email,
    required String password,
    String? username,
  }) async {
    final deviceId = await _deviceIdStore.getOrCreate();
    final body = <String, dynamic>{
      'email': email,
      'password': password,
      'deviceId': deviceId,
    };
    if (username != null && username.isNotEmpty) {
      body['username'] = username;
    }

    final data = await _client.post('/auth/register', body: body);

    final session = AuthSession.fromJson(data);
    await _saveSession(session);
    return session;
  }

  Future<bool> isUsernameAvailable(String username) async {
    final data = await _client.get(
      '/users/username-available',
      query: {'username': username},
    );
    if (data is Map<String, dynamic>) {
      return data['available'] == true;
    }
    return false;
  }

  Future<void> requestPasswordReset({
    required String email,
  }) async {
    await _client.post('/auth/password-reset/request', body: {
      'email': email,
    });
  }

  Future<void> confirmPasswordReset({
    required String token,
    required String newPassword,
  }) async {
    await _client.post('/auth/password-reset/confirm', body: {
      'token': token,
      'newPassword': newPassword,
    });
  }

  Future<void> requestEmailOtp({
    required String email,
  }) async {
    await _client.post('/auth/email-otp/request', body: {
      'email': email,
    });
  }

  Future<void> verifyEmailOtp({
    required String email,
    required String code,
  }) async {
    await _client.post('/auth/email-otp/verify', body: {
      'email': email,
      'code': code,
    });
  }

  Future<void> updateUserDetails({
    required String firstName,
    required String lastName,
    required String phoneCountryCode,
    required String phoneNumber,
  }) async {
    await _client.put(
      '/users/me/details',
      body: {
        'firstName': firstName,
        'lastName': lastName,
        'phoneCountryCode': phoneCountryCode,
        'phoneNumber': phoneNumber,
      },
      auth: true,
    );
  }

  Future<void> _saveSession(AuthSession session) async {
    if (session.accessToken.isNotEmpty) {
      await _store.setAccessToken(session.accessToken);
    }
    if (session.refreshToken.isNotEmpty) {
      await _store.setRefreshToken(session.refreshToken);
    }
    if (session.userId.isNotEmpty) {
      await _store.setUserId(session.userId);
    }
  }
}

File: apps/web/src/services/auth-service.ts
/**
 * Auth Service
 * Matches mobile's auth_service.dart
 */

import { apiClient } from '../adapters/api-client';
import { secureStore } from '../adapters/secure-store';
import { getOrCreateDeviceId, getDeviceName, getPlatform } from '../adapters/device-id';

export interface AuthSession {
  userId: string;
  email: string;
  accessToken: string;
  refreshToken: string;
  isVerified: boolean;
}

interface AuthResponse {
  accessToken: string;
  refreshToken: string;
  user: {
    id: string;
    email: string;
    isVerified: boolean;
    username?: string;
  };
}

interface UsernameAvailableResponse {
  available: boolean;
}

class AuthService {
  async login(email: string, password: string): Promise<AuthSession> {
    const deviceId = getOrCreateDeviceId();
    const deviceName = getDeviceName();
    const platform = getPlatform();

    const data = await apiClient.post<AuthResponse>('/auth/login', {
      body: {
        email,
        password,
        deviceId,
        deviceName,
        platform,
      },
    });

    const session = this.parseSession(data);
    await this.saveSession(session);
    return session;
  }

  async register(email: string, password: string, username?: string): Promise<AuthSession> {
    const deviceId = getOrCreateDeviceId();
    const deviceName = getDeviceName();
    const platform = getPlatform();

    const body: Record<string, string> = {
      email,
      password,
      deviceId,
      deviceName,
      platform,
    };

    if (username) {
      body.username = username;
    }

    const data = await apiClient.post<AuthResponse>('/auth/register', { body });

    const session = this.parseSession(data);
    await this.saveSession(session);
    return session;
  }

  async isUsernameAvailable(username: string): Promise<boolean> {
    const data = await apiClient.get<UsernameAvailableResponse>('/users/username-available', {
      query: { username },
    });
    return data.available === true;
  }

  async requestEmailOtp(email: string): Promise<void> {
    await apiClient.post('/auth/email-otp/request', {
      body: { email },
    });
  }

  async verifyEmailOtp(email: string, code: string): Promise<void> {
    await apiClient.post('/auth/email-otp/verify', {
      body: { email, code },
    });
  }

  async requestPasswordReset(email: string): Promise<void> {
    await apiClient.post('/auth/password-reset/request', {
      body: { email },
    });
  }

  async confirmPasswordReset(token: string, newPassword: string): Promise<void> {
    await apiClient.post('/auth/password-reset/confirm', {
      body: { token, newPassword },
    });
  }

  async updateUserDetails(details: {
    firstName: string;
    lastName: string;
    phoneCountryCode: string;
    phoneNumber: string;
  }): Promise<void> {
    await apiClient.put('/users/me/details', {
      body: details,
      auth: true,
    });
  }

  async logout(): Promise<void> {
    const deviceId = getOrCreateDeviceId();
    try {
      await apiClient.post('/auth/logout', {
        auth: true,
        body: { deviceId },
      });
    } catch {
      // Ignore logout errors
    }
    secureStore.clearSession();
  }

  isLoggedIn(): boolean {
    return !!secureStore.getAccessToken();
  }

  getUserId(): string | null {
    return secureStore.getUserId();
  }

  private parseSession(data: AuthResponse): AuthSession {
    return {
      userId: data.user?.id || '',
      email: data.user?.email || '',
      accessToken: data.accessToken || '',
      refreshToken: data.refreshToken || '',
      isVerified: data.user?.isVerified === true,
    };
  }

  private async saveSession(session: AuthSession): Promise<void> {
    if (session.accessToken) {
      secureStore.setAccessToken(session.accessToken);
    }
    if (session.refreshToken) {
      secureStore.setRefreshToken(session.refreshToken);
    }
    if (session.userId) {
      secureStore.setUserId(session.userId);
    }
  }
}

export const authService = new AuthService();
